Major List refactor with few breaking changes

mlms13 opened the pull request:
There has been a lot of moving-things-around:

- Typeclass implementations now live in Relude_List_Types
- Free functions from typeclasses live in that same file
- List-specific functions live in Relude_List_Base (which can use `_Types`)
- `ListF` as been renamed to `List_Submodules`
- Everything list-related now lives in `src/data/list`
- There's an rei file that kept me honest during these changes

A few things broke (not really due to the refactor... more because I noticed some old naming and changed it):

- `List.containsF` is now `List.containsBy`
- `List.indexOfF` is now `List.indexOfBy`
- `List.sortF` is now `List.sortBy`
- `List.sumFloat` is now `List.Float.sum` (and there are lots of other List.Float helpers)
- `List.Sequence` no longer exists (it's created on the fly in `NonEmpty`)

Overall, these changes should make it quite a bit easier to avoid circular dependencies. It's also easier to add new functions to e.g. `Relude_Foldables` and automatically have them show up in `Relude.List`. If new functions depend on features of the inner type (like `Ord`), those functions will show up in `List.Int` and `List.Float` and friends with a lot less code-changing.

Hopefully this will make it quite a bit easier to have `Relude.Array` reuse a lot of the same work (#27) and add special Foldable functions like `min` and `max` (#30) to both Array and List.

andywhite37 commented:
I haven't taken a close look at this, but I'm all for merging it when you feel comfortable so we can forge ahead.  Also, if these same re-orgs makes sense for `Array`, I'm :+1: to make the parallel changes.

We can haggle over names of things sometime later :smile_cat:

mlms13 commented:
I feel pretty good about where the `List` stuff ended up, and almost all of it was doable without breaking changes. I'm about 70% through Array, which I kind o want to merge along with this, so I'll try to get them both into master tomorrow. :+1:

mlms13 commented:
Ok, I finished up making `Array` match `List`... here are more things that changed (so I can find this easier when I try to write release notes):

Removed or renamed:

- Remove `Array.one` (use `Array.pure` instead)
- `Array.contains` -> `Array.containsBy`
- `Array.indexOf` -> `Array.indexOfBy`
- `Array.sort` -> `Array.sortBy`
- `Array.distinct` -> `Array.distinctBy`
- Remove nested submodules in List (`Relude.List.Option.Traversable` and `Relude.List.Result.Traversable`)

Added:

- Add `List.Result.sequence`
- New `Array.sort` takes an ORD module
- New `Array.contains` takes an EQ module
- New `Array.indexOf` takes an EQ module
- New `Array.distinct` takes an EQ module (with a specialized string version in `Array.String`
- Add `Array.shuffle` (clones input array)
- Add `Array.remove`, `Array.removeBy`, `Array.removeEach`, `Array.removeEachBy`
- Add Array submodule helpers to match List (`Array.String`, `Array.Int`, `Array.Float`, `Array.Option`, and `Array.Result`)
